### PR TITLE
add json serializable type

### PR DESCRIPTION
**Pull request by gen740**
## Motivation

Setting a non json-serializable type to `system_attr` will cause a runtime
error. To avoid this runtime error, add `JSONSerializable` to the type checking
for mypy.

## Description of the changes

- Create `optuna/_typing.py` and add `JSONSerializable`.
- Use `JSONSerializable` to the `set_system_attr` function.

### CODE REVIEW

**Comment by contramundum53** [comment]
@HideakiImamura Could you review this?

**Comment by HideakiImamura** [comment]
@contramundum53 Could you review this PR?

File: optuna/integration/chainermn.py [hunk 1]
```diff
@@ -251,7 +252,7 @@ def set_user_attr(self, key: str, value: Any) -> None:
         self.comm.mpi_comm.barrier()
 
     @deprecated_func("3.1.0", "6.0.0")
-    def set_system_attr(self, key: str, value: Any) -> None:
+    def set_system_attr(self, key: str, value: JSONSerializable) -> None:
```
Comment (HideakiImamura) [inline optuna/integration/chainermn.py:255]: This function is deprecated. We don't have to change this.

File: optuna/integration/pytorch_distributed.py [hunk 2]
```diff
@@ -274,7 +275,7 @@ def set_user_attr(self, key: str, value: Any) -> None:
 
     @broadcast_properties
     @deprecated_func("3.1.0", "6.0.0")
-    def set_system_attr(self, key: str, value: Any) -> None:
+    def set_system_attr(self, key: str, value: JSONSerializable) -> None:
```
Comment (HideakiImamura) [inline optuna/integration/pytorch_distributed.py:278]: Ditto.

File: optuna/multi_objective/study.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 from optuna import multi_objective
 from optuna._deprecated import deprecated_class
 from optuna._deprecated import deprecated_func
+from optuna._typing import JSONSerializable
```
Comment (HideakiImamura) [inline optuna/multi_objective/study.py:18]: This module itself is deprecated. So, ditto.

File: optuna/multi_objective/trial.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
 
 from optuna import multi_objective
 from optuna._deprecated import deprecated_class
+from optuna._typing import JSONSerializable
```
Comment (HideakiImamura) [inline optuna/multi_objective/trial.py:12]: Ditto.

File: optuna/study/_frozen.py [hunk 5]
```diff
@@ -40,7 +40,7 @@ def __init__(
         study_name: str,
         direction: Optional[StudyDirection],
         user_attrs: Dict[str, Any],
-        system_attrs: Dict[str, Any],
+        system_attrs: Dict[str, Any],  # TODO(gen740): Change Any to JSONSerializable
```
Comment (HideakiImamura) [inline optuna/study/_frozen.py:43]: This is my personal opinion, but the TODO comment to leave in the code should be self-explanatory as to whether or not it should be done. Whether or not to make the type JsonSerializable for an API that is open to users is a matter of debate and needs to be agreed upon by the developers. Wouldn't it be enough, for example, to leave such a TODO discussion on the PR like this now?

File: optuna/study/_study_summary.py [hunk 6]
```diff
@@ -56,7 +56,7 @@ def __init__(
         direction: Optional[StudyDirection],
         best_trial: Optional[trial.FrozenTrial],
         user_attrs: Dict[str, Any],
-        system_attrs: Dict[str, Any],
+        system_attrs: Dict[str, Any],  # TODO(gen740): Change Any to JSONSerializable
```
Comment (HideakiImamura) [inline optuna/study/_study_summary.py:59]: Ditto.

File: optuna/study/study.py [hunk 7]
```diff
@@ -657,7 +659,7 @@ def objective(trial):
         self._storage.set_study_user_attr(self._study_id, key, value)
 
     @deprecated_func("3.1.0", "6.0.0")
-    def set_system_attr(self, key: str, value: Any) -> None:
+    def set_system_attr(self, key: str, value: JSONSerializable) -> None:
```
Comment (HideakiImamura) [inline optuna/study/study.py:662]: This is deprecated, so ditto.

File: optuna/trial/_base.py [hunk 8]
```diff
@@ -100,7 +101,7 @@ def set_user_attr(self, key: str, value: Any) -> None:
 
     @abc.abstractmethod
     @deprecated_func("3.1.0", "6.0.0")
-    def set_system_attr(self, key: str, value: Any) -> None:
+    def set_system_attr(self, key: str, value: JSONSerializable) -> None:
```
Comment (HideakiImamura) [inline optuna/trial/_base.py:104]: ditto.

File: optuna/study/study.py [hunk 9]
```diff
@@ -773,7 +775,7 @@ def objective(trial):
 
     def enqueue_trial(
         self,
-        params: Dict[str, Any],
+        params: Dict[str, Any],  # TODO(gen740): Change Any to JSONSerializable
```
Comment (HideakiImamura) [inline optuna/study/study.py:778]: ditto.

File: optuna/trial/_fixed.py [hunk 10]
```diff
@@ -133,7 +134,7 @@ def set_user_attr(self, key: str, value: Any) -> None:
         self._user_attrs[key] = value
 
     @deprecated_func("3.1.0", "6.0.0")
-    def set_system_attr(self, key: str, value: Any) -> None:
+    def set_system_attr(self, key: str, value: JSONSerializable) -> None:
```
Comment (HideakiImamura) [inline optuna/trial/_fixed.py:137]: ditto.

File: optuna/trial/_frozen.py [hunk 11]
```diff
@@ -142,7 +145,7 @@ def __init__(
         params: Dict[str, Any],
         distributions: Dict[str, BaseDistribution],
         user_attrs: Dict[str, Any],
-        system_attrs: Dict[str, Any],
+        system_attrs: Dict[str, Any],  # TODO(gen740): Change Any to JSONSerializable
```
Comment (HideakiImamura) [inline optuna/trial/_frozen.py:148]: ditto.

File: optuna/trial/_frozen.py [hunk 12]
```diff
@@ -295,7 +298,7 @@ def set_user_attr(self, key: str, value: Any) -> None:
         self._user_attrs[key] = value
 
     @deprecated_func("3.1.0", "6.0.0")
-    def set_system_attr(self, key: str, value: Any) -> None:
+    def set_system_attr(self, key: str, value: JSONSerializable) -> None:
```
Comment (HideakiImamura) [inline optuna/trial/_frozen.py:301]: ditto.

File: optuna/trial/_trial.py [hunk 13]
```diff
@@ -585,7 +586,7 @@ def objective(trial):
         self._cached_frozen_trial.user_attrs[key] = value
 
     @deprecated_func("3.1.0", "6.0.0")
-    def set_system_attr(self, key: str, value: Any) -> None:
+    def set_system_attr(self, key: str, value: JSONSerializable) -> None:
```
Comment (HideakiImamura) [inline optuna/trial/_trial.py:589]: ditto.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR and sorry foe the late review. I have several comments. Please take a look.

**Review by HideakiImamura — APPROVED** [review]
LGTM!

**Review by contramundum53 — APPROVED** [review]
LGTM.
Notes: In this PR, the arguments of setters of `system_attrs` are typed as `Mapping[str, JSONSerializable]`, while the getters and internal fields are typed as `Dict[str, Any]`. This is reasonable because we do not expect users to JSON-serialize the `system_attr`s.

**Pull request merged**